opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/services/DiscoveryServices.java

```diff
@@ -0,0 +1,345 @@
+import org.eclipse.milo.opcua.stack.core.UaException;
+import org.eclipse.milo.opcua.stack.core.application.services.ServiceRequest;
+import org.eclipse.milo.opcua.stack.core.types.builtin.DateTime;
+import org.eclipse.milo.opcua.stack.core.types.builtin.DiagnosticInfo;
+import org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject;
+import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
+import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
+import org.eclipse.milo.opcua.stack.core.types.structured.FindServersOnNetworkRequest;
+import org.eclipse.milo.opcua.stack.core.types.structured.FindServersOnNetworkResponse;
+import org.eclipse.milo.opcua.stack.core.types.structured.MdnsDiscoveryConfiguration;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServer2Request;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServer2Response;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServerRequest;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisterServerResponse;
+import org.eclipse.milo.opcua.stack.core.types.structured.RegisteredServer;
+import org.eclipse.milo.opcua.stack.core.types.structured.ResponseHeader;
+import org.eclipse.milo.opcua.stack.core.types.structured.ServerOnNetwork;
+import org.eclipse.milo.opcua.stack.server.tcp.DefaultDiscoveryService;
+import org.eclipse.milo.opcua.stack.server.tcp.UaTcpStackServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DiscoveryServices extends DefaultDiscoveryService {
+
+    private LinkedList<RegisteredServer> registeredServers;
+    private HashMap<RegisteredServer, Date> registeredServerLastSeen;
+
+    private final boolean multicastEnabled;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private MdnsHelper mdnsHelper = null;
+
+
+    private Consumer<RegisteredServer> registerServerConsumer = null;
+
+    public DiscoveryServices(UaTcpStackServer server, boolean multicastEnabled) {
+        super(server);
+        registeredServers = new LinkedList<>();
+        registeredServerLastSeen = new HashMap<>();
+        this.multicastEnabled = multicastEnabled;
+        new Timer().scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                checkRegistrationTimeout();
+            }
+        }, 30 * 1000, 30 * 1000); // check cleanup every 30 secs
+
+        if (multicastEnabled) {
+            OpcUaServerConfig config = (OpcUaServerConfig) server.getConfig();
+
+            Set<String> hostnames = new HashSet<>();
+            hostnames.add(config.getHostname());
+
+            for (String bindAddress : config.getBindAddresses()) {
+                hostnames.addAll(config.getHostnameResolver().apply(bindAddress));
+            }
+
+            mdnsHelper = new MdnsHelper(hostnames);
+            Thread mdnsThread = new Thread(mdnsHelper);
```

Comment:
Same concern, it looks like `MdnsHelper` sleeps forever, how do we clean this up?